pySDC/helpers/problem_helper.py

```diff
@@ -130,3 +145,33 @@ def get_finite_difference_matrix(
     A /= dx**derivative
 
     return A
+
+
+def get_1d_grid(size, bc, left_boundary=0.0, right_boundary=1.0):
+    """
+    Generate a grid in one dimension and obtain mesh spacing for finite difference discretization.
+
+    Args:
+        size (int): Number of degrees of freedom per dimension
+        bc (str): Boundary conditions for both sides
+        left_boundary (float): x value at the left boundary
+        right_boundary (float): x value at the right boundary
+
+    Returns:
+        float: mesh spacing
+        numpy.ndarray: 1d mesh
+    """
+    L = right_boundary - left_boundary
+    if bc == 'periodic':
+        dx = L / size
+        xvalues = np.array([left_boundary + dx * i for i in range(size)])
+    elif "dirichlet" in bc:
+        dx = L / (size + 1)
+        xvalues = np.array([left_boundary + dx * (i + 1) for i in range(size)])
+    elif "neumann" in bc:
```

Comment:
I copied this from the generic ND FD class. I believe this may work for the special case of Dirichlet-zero. But really, I have no idea. I can remove this from here if you think it's fishy.

Reply:
No no, that's a good idea to have this here. But I just don't understand why you have this for neumann :

```python
dx = L / (size - 1)
```

and not this (same as dirichlet) :

```python
dx = L / (size + 1)
```

In both cases, you don't include the boundary points, what changes is simply how you compute the gradient at the points close to the boundary ...

Reply:
PS : you may want to write a short markdown notice somewhere on how the boundary are computed for neuman or dirichlet, with some formulas :wink:

Reply:
No, for Neumann, the boundaries are included, but for Dirichlet they are excluded. This is precisely the difference.

Reply:
I don't get it ... let say you are using first derivative with first order finite difference, on the left boundary :

$$
v = \frac{\partial u}{\partial x} (x_0) \simeq \frac{u_1-u_0}{\delta x}
$$

then if you know $v$ and $u_1$ (first interior point), then the value for $u_0$ is already determined and not an unknown anymore, right ?

Reply:
Yes, but don't you have to get $u_1$ and $u_0$ simultaneously? Where do you know $u_1$ from? That's why I put the first derivative in the FD matrix, so you solve for both at the same time...? 🤷‍♂️

Reply:
Oh, I see ... in my way, you have to include a constant $b$ vector when evaluating the derivative on your mesh : you don't solve an homogeneous problem anymore, but rather this one :

$$
\frac{du}{dt} = Au + b
$$

where $b$ is mostly zeros, except on the extremities. See more details below ... 

---

For instance, if you use first order Neumann boundary conditions like before :

$$
v = \frac{\partial u}{\partial x} (x_0) \simeq \frac{u_1-u_0}{\delta x}
$$

and centered finite difference of second order :

$$
\frac{\partial u}{\partial x} (x_1) \simeq \frac{u_2-u_0}{2\delta x}
$$

then you can compute the derivative in $x_1$ by replacing the Neumann condition in there :

$$
\frac{\partial u}{\partial x} (x_1) \simeq \frac{u_2-u_1}{2\delta x} + \frac{v}{2}
$$

Then you just have $b_1 = \frac{v}{2}$ and have to adapt the $A$ matrix so it computes the previous formula in $x_1$. Note that this is the same idea when you use Dirichlet BC with non-zero boundary values ...

---

In your way, it seems like you avoid this $b$ term by adding $x_0$ in $A$ and when you compute the derivative, you simply put the $v$ value in there. I do wonder if both approach are equivalent at then end ... especially since in you case, you are not solving such a system :

$$
\frac{du}{dt} = Au
$$

but rather this one

$$
\frac{du}{dt} = A(u)
$$

where the $A$ operator is not linear (for instance, $A(u + w)$ just put $v$ on the first value of its return vector, which is not the same as $A(u) + A(w)$ ...

So again : you should probably write the theory down somewhere, easier to check that first than to guess it from your code :wink:

Reply:
I have read about both approaches in the LeVeque FD book and I believe they work out similarly. But the second approach does not get rid of $b$.

I am solving 

$$
u_t = \Delta u + g(u),
$$

with $g$ some nonlinear term.
Then, I setup a grid $[0, dx, .. (N-1)dx, 1]$ for FD discretisation and write a matrix 

$$
A_{0,j} = (-1/dx, 1/dx, 0, ..., 0),\\
A_{N>I>0,j} = (0, ..., -1/dx^2, 2/dx^2, -1/dx^2, ...0),\\
A_{N,j} = (0, ..., 0, 1/dx, -1/dx).
$$

Which is first derivative at the boundaries and second derivative everywhere else. Actually, I use a one-sided stencil of the same order of accuracy as the centered second order derivative stencil, but I don't know the coefficients by heart and the concept doesn't change.

Then, I construct a vector $b$ such that $u_t \approx Au+b(u)$, with

$$
b_0 = b_N = v,\\
b_{N>j>0} = g(u).
$$

Now, the boundary will have an inhomogeneity according to the BCs and the inner part gets the non-linear term of the problem.

Since I only have $v=0$, I set the first and last entry of the "non-linear" term to 0 in the function that applies the BCs in the quench problem class. I can't think of a generic implementation in the problem helper for this. But is this valid in the first place? I went with this because it is the easiest to generalise to higher order. For instance, I get second order everywhere, whereas your approach is only first order, afaik. Only if both works, ofc.

I will ask at the hackathon in Darmstadt if anybody has a good test for this. That would be helpful.

Apologies for the missing line breaks...

Reply:
Actually, you can use second order boundary condition with my approach (I just don't have the formula in mind) ... you just have to use a shifted derivation stencil that evaluate the derivative that way :

$$
\frac{\partial u}{\partial x}(x_0) \simeq \frac{\alpha_0 u_0 + \alpha_1 u_1 + \alpha_2 u_2}{\delta x}
$$

then replace $u_0$ in the derivative formula on $x_1$ with a combination of $u_1$, $u_2$ and $v$. Then you don't have a non-linear $b$ term anymore.

Didn't know about the Leveque FD book (guess that's some cultural reference missing on my side), but I guess we may have to set on one approach to use for each non-periodic boundary conditions, so it can be consistent ? 

If I call them the non-linear approach (yours) of the linear approach (mine), then Dirichlet-zero BC uses linear while Neumann uses non-linear, and I did not check if the Dirichlet-nonZero BC was implemented ... Also, choosing one approach for both would allows to mix Dirichlet and Neumann type BC, which has to be duable in theory ...

Reply:
PS : after reading carefully, looking at your $A_{0,j}$ line, it seems that the derivative for $x_0$ is then

$$
\frac{u_1 - u_0}{\delta x} + v
$$

does not looks for me (and yes, it's the first order coefficients, normally you would use the second order shifted one, but still ...)

Reply:
There is probably a sign error. I didn't think about this because I only use zero. And I guess you have to implement both in the problem class. But the version in this PR is not intrinsically non-linear, the non-linearity comes from the problem itself. I guess the advantage of your approach is that it uses 2 fewer unknowns to cover the same interval. The advantage of my approach is that it is more easily generalised to high order. In fact, arbitrary order is already implemented. Your approach is a bit more difficult because you have to cancel some extra terms. This sounds doable, and I guess I can try if you want. Mixing BCs should be possible either way, I think. You just need a bit of extra care.

Is the sign of $v$ what worries you in the last post or is it deeper concerns? If it's the sign, I would suggest merging this and worrying only when someone needs sth more special.

Reply:
Not the sign of $v$ ... the formula looks false, because the derivative should be $v$ at the end :sweat_smile:

Reply:
As I said ... shouldn't be like this. This won't probably change the result that much in the end, especially if you have a lot of points in space. But still : wrong formula. If you are OK with approximating stuff if it's simpler, then go for it.

If not, I would suggest trying the "linear" approach. In both case, you'll need some shifted differentiation stencil (that can be retrieved for uniform mesh here : https://en.wikipedia.org/wiki/Finite_difference_coefficient#Arbitrary_stencil_points). Then it's just a small manipulation to generalize for high order, for both non-linear and linear formulation (in particular, the A matrix should be adapted close to the boundary when using stencils with many points, e.g for order 4).

If you don't have (or will) to look at this on your own, we can probably implement this in one or two hours in Darmstadt :wink: